content/en/docs/components/pipelines/author-a-pipelines/bring_your_own_container.md

```diff
@@ -0,0 +1,184 @@
++++
+title = "Building components using your own container"
+description = "Bring user-controlled container into component authoring in KFP v2"
+weight = 50
++++
+
+Container-based components make it easier to use the pre-existed container to build a KFP component. 
+The user will control the container entrypoint and is able to define the component by specifying their container, commands, and args in a `ContainerSpec` object with similar authoring experiences as defining Python components. 
+
+
+## Before you begin
+
+1. Run the following command to install the Kubeflow Pipelines SDK v(TODO) or higher. If you run this command in a Jupyter notebook, restart the kernel after installing the SDK. 
+
+```python
+$ pip install --upgrade kfp
+```
+
+2. Import the `kfp`, and `kfp.dsl` packages.
+
+```python
+import kfp
+from kfp.dsl import (
+  container_component,
+  ContainerSpec,
+  pipeline,
+  OutputPath
+)
+ 
+```
+
+3. Create an instance of the [`kfp.Client` class][kfp-client] following steps in [connecting to Kubeflow Pipelines using the SDK client][connect-api].
+
+
+```python
+client = kfp.Client() # change arguments accordingly
+```
+
+For more information about the Kubeflow Pipelines SDK, see the [SDK reference guide][sdk-ref].
+
+
+## Getting started with Container-based components
+
+1. This section demonstrates how to get started building container-based components by creating a simple hello world component.
+
+Define your component’s code as a python function that returns a `dsl.ContainerSpec` object to specify the container image and the commands to be run in the container and wrap the function into a `@container_component` decorator. The decorator will compose a `ComponentSpec` object using the `ContainerSpec` to be compiled for execution. (Learn more about [ComponentSpec][dsl-component-spec] and [ContainerSpec][dsl-container-spec].) The inputs and outputs of the component should be written in the function signature with type annotations. In this example, we use a shell script to demonstrate how you could use a simple program in the container to pass data through this interface.  
+  <!-- TODO: add hyperlink to source code of ContainerSpec  -->
+
+```python
+@container_component
+def hello_world_comp(text: str, out_filename: dsl.OutputPath(str)):
+    return ContainerSpec(
+        image=..., # put your container image on gcs here
```

Comment:
Container registry is GCR rather than GCS. BTW, what if we reference an image from docker hub. Would that work on Vertex?

Would it be more straightforward to put an actual image here? WDYT?

Reply:
Yes, I think I should change it to an actual example now (e.g. python:3.7)